pandas/core/range.py

```diff
@@ -0,0 +1,427 @@
+import pandas as pd
+import numpy as np
+import pandas.algos as _algos
+from pandas.core.index import Int64Index, Index
+import pandas.core.index as pdindex
+
+EMPTY_RANGE = lambda: pd.Index([], dtype='int64')
+
+def _delegate_to_int64(func_name, doc=None):
+    def wrapper(self, *args, **kwargs):
+        return getattr(self.as_int64index, func_name)(*args, **kwargs)
+    wrapper.__name__ = func_name
+    if hasattr(Int64Index, func_name):
+        doc = doc or getattr(Int64Index, func_name).__doc__
+    wrapper.__doc__ = doc or ''
+    return wrapper
+
+def _not_implemented(self, *args, **kwargs):
+    raise NotImplementedError('method not implemented')
+
+class RangeIndex(Index):
+    """Represents a range with left-open interval.
+
+    Parameters
+    ----------
+    left, right : int
+        start and end of range (i.e., ``s[left:right]``). If left > right,
+        assumes reversed (i.e., ``s[left:right:-1]``)
+    """
+    _groupby = _algos.groupby_range
+    # My thinking on this structure:
+    # - from a 'set theoretic' standpoint, order doesn't matter, so ascending
+    #   vs. descending is generally an implementation detail
+    # - left, right are necessary for comparison, but making start and end
+    #   separate makes it much easier to work with
+    # - prohibiting empty RangeIndex helps simplify edge cases considerably.
+    # - From pandas' perspective, RangeIndex should behave *exactly* the same
+    #   as an Int64Index. (except for ops where it can yield more RangeIndexes)
+    # - supporting steps might be possible, but less simple (and less clear
+    #   that it would be useful to pandas proper, given that you'd have to know
+    #   ahead of time that you could convert to a stepped range). Plus, only
+    #   helpful when you specifically have a consistent step
+    # - Certain operations with RangeIndex could benefit from allowing nested
+    #   ranges, i.e. the union of RangeIndex(10, 5) and RangeIndex(3, 7) could be
+    #   something like: [RangeIndex(10,5), RangeIndex(3, 7)] and then could just
+    #   iterate over that. But that's for after all of this starts working.
+    # - It would be nice if groupby() accepted an Index
+    # - It might be valuable to cache the values property of a RangeIndex, but
+    #   I'm not totally convinced that's the best strategy (yet!), unless RangeIndex
+    #   could *replace* itself on its parent. (tradeoff between instantiation time and
+    #   ability to gc values when they aren't needed anymore)
+    # TODO: Block setting of start and end
+    def __new__(cls, left, right=None, step=1, name=None):
```

Comment:
I would follow the signature of the builtin range here:
```
class range(stop, *, name=None)
class range(start, stop, step=1, *, name=None)
```

You'll need to parse `args` and `kwargs` manually to make this work.